File: src/main/java/com/gft/webinar/kip/_6_collections/Grouping.java
package com.gft.webinar.kip._6_collections;

import com.gft.webinar.kip.Department;
import com.gft.webinar.kip.Employee;
import com.gft.webinar.kip.util.EmployeesProvider;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.function.BinaryOperator;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Grouping {

    public static void main(String[] args) {
        highestPaidEmployees(EmployeesProvider.INSTANCE.getEmployees()).forEach(((department, employee) -> {
            System.out.println("""
                %s: %s %s - %s"""
                .stripIndent()
                .formatted(department.getName(), employee.getFirstName(), employee.getLastName(), employee.getSalary())
            );
        }));

        System.out.println(totalSalaryPerDepartment(EmployeesProvider.INSTANCE.getEmployees()));
    }

    public static Map<Department, Employee> highestPaidEmployees(List<Employee> employees) {
        return employees.stream()
            .collect(Collectors.toMap(
                Employee::getDepartment,
                Function.identity(),
                BinaryOperator.maxBy(Comparator.comparingDouble(Employee::getSalary))
            ));
    }

    public static Map<Department, Double> totalSalaryPerDepartment(List<Employee> employees) {
        return employees.stream()
            .collect(Collectors.groupingBy(
                Employee::getDepartment,
                Collectors.summingDouble(Employee::getSalary)
            ));
    }

}

File: src/main/kotlin/com/gft/webinar/kip/_6_collections/Grouping.kt
package com.gft.webinar.kip._6_collections

import com.gft.webinar.kip.Department
import com.gft.webinar.kip.Employee
import com.gft.webinar.kip.util.EmployeesProvider

/*
 * Objectives:
 *
 * 1. GroupBy vs GroupingBy + the Grouping type
 *
 *  -> Find the highest paid employees per department
 *  -> What is Grouping and how can we benefit from it
 */

fun main() {

	val employees = EmployeesProvider.employees
	println("Highest paid employees per department:")
	highestPaidEmployees(employees).forEach {
		println("\t${it.key.name}: ${it.value?.firstName} ${it.value?.lastName} (${it.value?.gender}) - ${it.value?.salary}")
	}

	// ========== Grouping

	val grouping = employees.groupingBy { employee -> employee.department }

	// what can we do with it?

	// 1. how many employees there are in each department
	println("Employees per department count: ")
	grouping.eachCount().forEach {
		println("\t${it.key.name}: ${it.value}")
	}

	// 2. total salary per department
	println("Total salary per department:")
	grouping
		.fold(0.0) { accumulator, employee ->
			accumulator + employee.salary
		}
		.forEach { department, totalSalary ->
			println("\t${department.name}: $totalSalary")
		}

}

/**
 * This function must return the highest paid employee in each department.
 */
fun highestPaidEmployees(employees: List<Employee>): Map<Department, Employee?> {
	return employees
		.groupBy { employee -> employee.department }
		.mapValues { entry ->
			entry.value.maxBy { employee -> employee.salary }
		}
}